### PR TITLE
MLA perf

### PR DESCRIPTION
**Pull request by quic-mamta**
1. Added support for Kimi-K2 models with DeepSeek modeling.
2. Added MLA with/WO fusion
3. Implemented Dual Cache for compressed KV and k_rope
4. Introduced prefill-only MoE changes.
5. Implemented KV blocking and head blocking strategies.
6. Added replicate KV head transform for kimi-k2
7. Performance improvements for Multi-head LatentAttention (MLA)
8. Add support for subfunctions
9. Use generalized kv and h blocking for MLA

### CODE REVIEW

File: examples/kimi_k2/export_kimik2.py
Comment (ochougul) [inline examples/kimi_k2/export_kimik2.py:1]: remove this file

File: examples/text_generation/run_kimik2.py
Comment (ochougul) [inline examples/text_generation/run_kimik2.py:1]: Update to use qaic_config only. Let's remove env variable

File: examples/kimi_k2/run_orig_kimi_k2.py
Comment (ochougul) [inline examples/kimi_k2/run_orig_kimi_k2.py:1]: REmove this, this is for developer purpose

File: QEfficient/blocking/get_num_blocks.py
Comment (ochougul) [inline QEfficient/blocking/get_num_blocks.py:1]: Don't create this new file, add this code in `blocking_configurator.py`

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek_orig.py
Comment (ochougul) [inline QEfficient/transformers/models/deepseek_v3/modeling_deepseek_orig.py:1]: remove this file.

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py
Comment (ochougul) [inline QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py:1]: remove this file.

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek_qeff.py
Comment (ochougul) [inline QEfficient/transformers/models/deepseek_v3/modeling_deepseek_qeff.py:1]: rename this to `modeling_deepseek.py`

File: examples/kimi_k2/export_kimik2.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+from QEfficient import QEFFAutoModelForCausalLM
+
+# parameters to be configured
+
+prompt = "Once upon a time,"
+num_kv_heads_repeat = 1  # When using KIMI_BLOCKING="kv", make sure num_kv_heads_repeat is set to 1. Use only for KIMI_BLOCKING="h" when using MLA and this should be equal to TS in that case.
+num_hidden_layers = 2
+TS = 4
+enable_mla = True
+mla_absorption_config = {"enable": False, "online": False}
```
Comment (ochougul) [inline examples/kimi_k2/export_kimik2.py]: Lets change this to
cache_compressed: True/False 
absorption: True/False -> gets enabled only when cache_compressed is True
online: True/False -> gets enabled only when absorption is True

**Review by ochougul — CHANGES_REQUESTED** [review]

**Review by vbaddi — APPROVED** [review]

File: examples/text_generation/run_kimik2.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+# ----------------------------------------------------------------------------
+
+import numpy as np
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from QEfficient import QEFFAutoModelForCausalLM
+
+# parameters to be configured
+prompt = "Once upon a time,"
```
Comment (quic-rishinr) [inline examples/text_generation/run_kimik2.py:15]: Please move the example under text_generation/models/kimi and follow the structure we are following for both example scripts. Also create a readme and and inline comments.

File: examples/text_generation/run_kimik2.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+# with torch.no_grad():
+#    out = model(**inputs)
+#    predictions = torch.argmax(out.logits, dim=-1)
+
+qeff_model = QEFFAutoModelForCausalLM(model)
+qeff_model.transform(ctx_len=CTX_LEN, seq_len=PREFILL_SEQ_LEN, bs=1, num_devices=TS, qaic_config=qaic_config)
+qeff_model.mla(mla_absorption_config=mla_absorption_config)
+
+inputs = tokenizer(prompt, return_tensors="np", padding="max_length", max_length=padded_len)
```
Comment (quic-rishinr) [inline examples/text_generation/run_kimik2.py:56]: Do we need the generation logic in examle script? better to have qeff_model.generate() in the example script to keep it simple

File: examples/kimi_k2/export_kimik2.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+# qaic_config = None #for orig_forward
+# qaic_config = {"num_kv_heads_repeat": TS}  #with head replication for orig_forward
+qaic_config = {
+    "enable_blocking": True,
+    "blocking_mode": "h",
+    "num_kv_heads_repeat": TS,
+}  # for h blocking, it internally sets head_block_size equal to num_devices/num_kv_heads_repeat
+# qaic_config = {"enable_blocking": True, "blocking_mode": "kv"}  # for KV blocking
+
+# model_path = "/home/ochougul/.cache/huggingface/hub/models--moonshotai--Kimi-K2-Thinking/snapshots/a51ccc050d73dab088bf7b0e2dd9b30ae85a4e55/"
```
Comment (quic-rishinr) [inline examples/kimi_k2/export_kimik2.py]: remove local model paths

File: QEfficient/transformers/models/modeling_auto.py [hunk 5]
```diff
@@ -2763,19 +2765,27 @@ class QEFFAutoModelForCausalLM(QEFFBaseModel):
 
     _onnx_transforms = []
 
+    def mla(
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py]: Why do we need to have the mla method in AutoModelClass? This would be irrelevant for most of the models. 
lets take it as a qaic_configs or compiler args or transforms. Remove this method

File: QEfficient/transformers/models/modeling_auto.py [hunk 6]
```diff
@@ -3074,10 +3084,26 @@ def export(
...
+            self.hash_params["mla_absorption_config"] = mla_absorption_config
+            setattr(self.model.model, "mla_absorption_config", mla_absorption_config)
+
+        if "DeepseekV3ForCausalLM" in (getattr(self.model.config, "architectures", None) or []):
+            if prefill_only:
+                self.prefill(enable=True)
+                self.hash_params["prefill_only"] = True
+            else:
+                self.prefill(enable=False)
+                self.hash_params.pop("prefill_only", None)
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py]: Can we handle this as an ConfigTransform?

File: QEfficient/transformers/models/modeling_auto.py [hunk 7]
```diff
@@ -3179,6 +3205,47 @@ def export(
                     dynamic_axes[f"past_{kv}.{i}"] = pkv_dynamic_axes[i]
                     output_names.append(f"past_{kv}.{i}_RetainedState")
 
+        if "DeepseekV3ForCausalLM" in (getattr(self.model.config, "architectures", None) or []):
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py:3196]: Can we move to the modelling file for generating dummy inputs? we should not be haivng this in the export method

File: QEfficient/transformers/models/deepseek_v3/configuration_deepseek.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    >>> from transformers import DeepseekV3Model, DeepseekV3Config
+
+    >>> # Initializing a Deepseek-V3 style configuration
+    >>> configuration = DeepseekV3Config()
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+
+    model_type = "deepseek_v3"
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/deepseek_v3/configuration_deepseek.py:123]: Do we need to maintain this configuration file? cant we read it from tf or config?

File: QEfficient/blocking/blocking_configurator.py [hunk 9]
```diff
@@ -20,6 +20,9 @@
 from QEfficient.utils import get_attr_or_key, require_value
 from QEfficient.utils.constants import VTCM_SIZE_THRESHOLD
 
+FP16_BYTES = 2
```
Comment (quic-rishinr) [inline QEfficient/blocking/blocking_configurator.py]: Please move it to Constants.py

File: QEfficient/blocking/blocking_configurator.py [hunk 10]
```diff
@@ -90,6 +93,92 @@ def block_candidates_generator(max_length: int) -> List[int]:
     return block_list
 
 
+def matmul1_bytes(q_len: int, kv_block_size: int, num_heads: int = DEFAULT_NUM_HEADS) -> int:
+    """Bytes for [1,num_heads,q,kv] x [1,1,kv,512] -> [1,num_heads,q,512] in fp16."""
+    elems_a = num_heads * q_len * kv_block_size
+    elems_b = kv_block_size * 512
```
Comment (quic-rishinr) [inline QEfficient/blocking/blocking_configurator.py]: declare magic number like 512 and 576 as constants in constants.py and reuse it here

File: QEfficient/base/modeling_qeff.py [hunk 11]
```diff
@@ -433,11 +449,22 @@ def transform(
 
         qaic_config = qaic_config if qaic_config else getattr(self.model, "qaic_config", None)
 
-        if getattr(self.model, "config", None) or getattr(self.model.model, "config", None):
+        model_config = getattr(self.model, "config", None) or getattr(self.model.model, "config", None)
+
+        if model_config:
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:454]: We should not be adding model specific changes in the base class. please remove it

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py [hunk 12]
```diff
@@ -0,0 +1,1160 @@
...
+                "beta_fast",
+                "beta_slow",
+                "mscale",
+                "mscale_all_dim",
+            ]
+            if key in self.config.rope_scaling
+        }
+        self.rotary_emb = DeepseekV3YarnRotaryEmbedding(
+            self.config.qk_rope_head_dim,
+            max_position_embeddings=32 * 1024,
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py]: Can we capture this in constants.py?

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py [hunk 13]
```diff
@@ -0,0 +1,1160 @@
...
+    if scale <= 1:
+        return 1.0
+    return 0.1 * mscale * math.log(scale) + 1.0
+
+
+def yarn_linear_ramp_mask(min, max, dim):
+    if min == max:
+        max += 0.001  # Prevent singularity
+
+    linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py:58]: Hardcoding the dtype will break the support for exporting the model in BF16/FP16 use dtype=module.config.torch_dtype similar to other modelling files

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py [hunk 14]
```diff
@@ -0,0 +1,1160 @@
...
+        self.per_head_k_up = torch.nn.Parameter(per_head_k_up.unsqueeze(0).detach().clone())
+        per_head_k_up_normal = self.per_head_k_up.transpose(2, 3)
+        self.per_head_k_up_normal = torch.nn.Parameter(per_head_k_up_normal.detach().clone())
+
+        fusedqk = torch.bmm(per_head_q_up, per_head_k_up).reshape(
+            -1, self.num_heads, self.q_lora_rank, self.kv_lora_rank
+        )
+        self.fusedqk = torch.nn.Parameter(fusedqk.detach().clone())
+
+    def fused_forward_h_blocking(
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py:254]: Can we combine or have a method for common changes in blocking? there are a lot of similar code between the h_blocking and kv_blocking

File: QEfficient/transformers/models/modeling_auto.py [hunk 15]
```diff
@@ -3355,9 +3422,11 @@ def compile(
         offload_pt_weights: Optional[bool] = True,
         enable_chunking: Optional[bool] = False,
         retain_full_kv: Optional[bool] = None,
+        mla_absorption_config: Optional[Dict[str, bool]] = None,
```
Comment (quic-rishinr) [inline QEfficient/transformers/models/modeling_auto.py]: Update the doc string for mla_absorption_config in both export and compile

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py (hunk 12 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py]: done

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py (hunk 14 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py:254]: there are still optimizations going on, we will restructure the modeling file , once changes are finalized.

File: QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py (hunk 13 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/deepseek_v3/modeling_deepseek.py:58]: this change was taken from original modeling file.

File: QEfficient/blocking/blocking_configurator.py (hunk 9 above)
Comment (quic-mamta) [reply QEfficient/blocking/blocking_configurator.py]: Done

File: QEfficient/blocking/blocking_configurator.py (hunk 10 above)
Comment (quic-mamta) [reply QEfficient/blocking/blocking_configurator.py]: Done

File: QEfficient/transformers/models/deepseek_v3/configuration_deepseek.py (hunk 8 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/deepseek_v3/configuration_deepseek.py:123]: this is an external modeling file downloaded from huggingface, it is not part of transformers.

File: examples/text_generation/run_kimik2.py (hunk 3 above)
Comment (quic-mamta) [reply examples/text_generation/run_kimik2.py:56]: qeff_model.generate won't work directly for this model, will move this code to API Runner once this architecture is officially supported, this is just to verify the pytorch runs.

File: QEfficient/transformers/models/modeling_auto.py (hunk 5 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/modeling_auto.py]: removed and moved to init, taking it from qaic_config.

File: examples/text_generation/run_kimik2.py (hunk 2 above)
Comment (quic-mamta) [reply examples/text_generation/run_kimik2.py:15]: Done

File: QEfficient/transformers/models/modeling_auto.py (hunk 7 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/modeling_auto.py:3196]: Done.

File: QEfficient/transformers/models/modeling_auto.py (hunk 6 above)
Comment (quic-mamta) [reply QEfficient/transformers/models/modeling_auto.py]: Done.

**Review by ochougul — APPROVED** [review]
We need to do 
- resolve the subfunction issue with head replication and forward_orig
-  check if with kv replication performance drops for KV blocking
- Code cleanup

Merging as of now
